core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -0,0 +1,379 @@
+
+   /**
+    *
+    * @param segment
+    * @param key
+    * @return
+    */
+   default CompletionStage<Boolean> containsKey(int segment, Object key) {
+      return loadEntry(segment, key)
+            .thenApply(Objects::nonNull);
+   }
+
+   /**
+    *
+    * @param segment
+    * @param entry
+    * @return
+    */
+   CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> entry);
+
+   /**
+    *
+    * @param segment
+    * @param key
+    * @return
+    */
+   CompletionStage<Boolean> delete(int segment, Object key);
+
+   /**
+    *
+    * @param segments
+    * @return
+    */
+   default CompletionStage<Void> addSegments(IntSet segments) {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.SEGMENTABLE + ", but it does not implement addSegments");
+   }
+
+   /**
+    *
+    * @param segments
+    * @return
+    */
+   default CompletionStage<Void> removeSegments(IntSet segments) {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.SEGMENTABLE + ", but it does not implement removeSegments");
+   }
+
+   /**
+    *
+    * @return
+    */
+   CompletionStage<Void> clear();
+
+   /**
+    *
+    * @param publisherCount
+    * @param publisher
+    * @return
+    */
+   default CompletionStage<Void> bulkWrite(int publisherCount, Publisher<SegmentedPublisher<MarshallableEntry<K, V>>> publisher) {
+      return Flowable.fromPublisher(publisher)
+            .concatMapCompletable(sp ->
+                        Flowable.fromPublisher(sp)
+                              .concatMapCompletable(me -> Completable.fromCompletionStage(write(sp.getSegment(), me)))
+                  , publisherCount)
+            .toCompletionStage(null);
+   }
+
+   /**
+    *
+    * @param publisherCount
+    * @param publisher
+    * @return
+    */
+   default CompletionStage<Void> bulkDelete(int publisherCount, Publisher<SegmentedPublisher<Object>> publisher) {
+      return Flowable.fromPublisher(publisher)
+            .concatMapCompletable(sp ->
+                        Flowable.fromPublisher(sp)
+                              .concatMapCompletable(obj -> Completable.fromCompletionStage(delete(sp.getSegment(), obj)))
+                  , publisherCount)
+            .toCompletionStage(null);
+   }
+
+   /**
+    *
+    * @param segments
+    * @return
+    */
+   default CompletionStage<Long> size(IntSet segments) {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.BULK_READ + ", but it does not implement size");
+   }
+
+   /**
+    *
+    * @param segments
+    * @return
+    */
+   default CompletionStage<Long> approximateSize(IntSet segments) {
+      return size(segments);
+   }
+
+   /**
+    *
+    * @param segments
+    * @param filter
+    * @return
+    */
+   default Publisher<MarshallableEntry<K, V>> entryPublisher(IntSet segments, Predicate<? super K> filter) {
```

Comment:
Actually looking again I thought I could remove the `fetchValue` argument. We will still want that. It is helpful for uses where we want the key and the metadata that maps to it but not the value. This is still beneficial. However there is no reason to not retrieve the metadata as far as I can see as it would not contain expiration info or other details.